backend/src/worker.rs

```diff
@@ -1351,175 +1360,261 @@ async fn handle_child(
+                .await
+                .err()
+                .map(|err| err.try_into_panic())
+            {
+                panic::resume_unwind(p);
             }
 
-            tokio::time::sleep(Duration::from_secs(5)).await;
-        }
-    });
+            (do_write, write_result) =
+                tokio::spawn(append_logs(job_id, joined, db.clone())).remote_handle();
 
-    let mut start = logs.chars().count();
-    let mut last_update = chrono::Utc::now().timestamp_millis();
-    let initial_start = chrono::Utc::now();
+            if let Err(err) = result {
+                tracing::error!(%job_id, %err, "error reading output for job {job_id}: {err}");
+                break;
+            }
 
-    while !done.load(Ordering::Relaxed) {
-        let diff = 500 - (chrono::Utc::now().timestamp_millis() - last_update);
-        let sleeping_future = if diff > 0 as i64 {
-            tokio::time::sleep(Duration::from_millis(diff as u64))
-        } else {
-            //TODO make it just resolve immediately
-            tokio::time::sleep(Duration::from_millis(0))
-        };
-        tokio::select! {
-            _ = sleeping_future => {
-                let end = logs.chars().count();
+            if *set_too_many_logs.borrow() {
+                break;
+            }
+        }
 
-                let to_send = logs.chars().skip(start).collect::<String>();
+        /* drop our end of the pipe */
+        drop(output);
 
-                if start != end {
-                    concat_logs(&to_send, &job_id, db).await;
-                    start = end;
-                }
+        if let Some(Ok(p)) = do_write
+            .then(|()| write_result)
+            .await
+            .err()
+            .map(|err| err.try_into_panic())
+        {
+            panic::resume_unwind(p);
+        }
+    };
 
-                let canceled = sqlx::query_scalar!("SELECT canceled FROM queue WHERE id = $1", job_id)
-                    .fetch_one(db)
+    /* a stream updating "queue"."last_ping" at an interval */
+
+    let ping = interval_skipping_missed(ping_interval)
+        .map(|_| db.clone())
+        .then(move |db| async move {
+            if let Err(err) =
+                sqlx::query!("UPDATE queue SET last_ping = now() WHERE id = $1", job_id)
+                    .execute(&db)
                     .await
-                    .map_err(|e| tracing::error!("error getting canceled for id {}: {e}", job_id))
-                    .unwrap_or(false);
+            {
+                tracing::error!(%job_id, %err, "error setting last ping for job {job_id}: {err}");
+            }
+        });
 
-                if canceled {
-                    tracing::info!("killed after cancel: {}", job_id);
-                    done.store(true, Ordering::Relaxed);
-                }
+    let wait_result = tokio::select! {
+        (w, _) = future::join(wait_on_child, lines) => w,
+        _ = ping.collect::<()>() => unreachable!("job ping stopped"),
```

Comment:
The unreachable part is weird, could you explain in more details ?
If really unreachable why even select! ?

Reply:
> The unreachable part is weird, could you explain in more details ?

The stream should repeat forever.

>  If really unreachable why even select! ?

To poll on the future.